design-pattern.md

```diff
@@ -183,3 +183,110 @@ public class Main {
+
+---
+
+## Pattern strutturali
+
+I pattern strutturali sono quelli che si occupano di come le classi e gli oggetti vengono composti per formare strutture più grandi.
+
+### Adapter
+
+L'**Adapter** è un pattern che permette di adattare un'interfaccia ad un'altra. Supponiamo di voler implementare una feature e, per farlo, ci serviamo di una libreria esterna. La libreria, però, lavora su dati e oggetti diversi da quelli con cui lavoriamo noi. In questo caso, possiamo creare un **Adapter** che si occupa di _adattare_ l'interfaccia della libreria esterna all'interfaccia che usiamo.
+
+#### UML
+
+![](./immagini/adapter.svg)
+
+In questo diagramma, abbiamo `Target` che definisce l'interfaccia che il Client usa, `Adaptee` che definisce l'interfaccia che vogliamo _adattare_ (ad esempio una libreria esterna), e `Adapter`, che estende `Target`, e usa `Adaptee` per adattare il methodo `methodToAdapt()` al metodo `request()` di `Target`.
+
+#### Esempio
+
+Supponiamo di voler implementare un programma che legge un file di testo e lo stampa a video. Per farlo, usiamo una libreria esterna, che però lavora su oggetti diversi da quelli che usiamo noi. In questo caso, possiamo creare un **Adapter** che si occupa di _adattare_ l'interfaccia della libreria esterna all'interfaccia che usiamo.
+
+Per prima cosa, definiamo l'interfaccia `Target`, in questo esempio la chiamiamo `TextPrinter`, che lavora con il tipo di file `TextFile`:
+
+```java
+public class TextFile {
+    private String name;
+    private String path;
+    private int size;
+
+    public TextFile
+    (String name, String path, int size) {
+        this.name = name;
+        this.path = path;
+        this.size = size;
+    }
+
+    public String getPath() {
+        return path;
+    }
+    
+    ...
+}
+
+public interface TextPrinter {
+    public void printText(TextFile textFile);
+}
+```
+
+Definiamo anche l'interfaccia `Adaptee` (la libreria esterna), in questo esempio la chiamiamo `FilePrinterLibrary`, che usera un tipo di file diverso da `TextFile` (`File`):
+
+```java
+public class File {
+    private String path;
+
+    public File(String path) {
+        this.path = path;
+    }
+
+    public String getPath() {
+        return path;
+    }
+
+    ...
+}
+
+public class FilePrinterLibrary {
+    public void printFile(File file) {
+        System.out.println(
+            "Printing file " + file.getPath());
+    }
+}
+```
+
+`TextPrinter` e `FilePrinterLibrary` lavorano con oggetti diversi, quindi, per adattare l'interfaccia di `FilePrinterLibrary` a quella di `TextPrinter`, creiamo un **Adapter**:
+
+```java
+public class FilePrinterLibraryAdapter
+implements TextPrinter {
+    @Override
+    public void printText(TextFile textFile) {
+        FilePrinterLibrary filePrinterLibrary =
```

Comment:
Credo convenga trasformare la `filePrinterLibrary` in un attributo, in modo da non doverla creare e distruggere ad ogni `printText`. Inoltre mettendola come attributo è più in linea con la composizione.
Puoi mettere nella classe
```java
public class ... {
    private FilePrinterLibrary filePrinterLibrary = new ...;
   
    ...
}
```